module/os/linux/zfs/zpl_file.c

```diff
@@ -422,8 +563,9 @@ zpl_aio_read(struct kiocb *kiocb, const struct iovec *iov,
 	crhold(cr);
 	cookie = spl_fstrans_mark();
 
+	int flags = (filp->f_flags | zfs_io_flags(kiocb)) & ~O_DIRECT;
 	int error = -zfs_read(ITOZ(filp->f_mapping->host), &uio,
-	    filp->f_flags | zfs_io_flags(kiocb), cr);
+	    flags, cr);
```

Comment:
I have not looked at the uncached I/O before. Feel free to point me at the code so I can get a better understanding of it.

Also, the point of clearing the `O_DIRECT` flag here is if we got an `EAGAIN` error when attempting to do a `O_DIRECT` write. This can happen with `direct=always` or in the case of the checksum verify thing. Happy to just send back `EIO` in the checksum verify error, but the `always` code would still rely on this at the moment.

Reply:
Please see https://github.com/openzfs/zfs/pull/14243 .  Right now it can be enabled only via primarycache=metadata, but I hoped to wire it to O_DIRECT also.

Reply:
This is an interesting idea. So this really is just handling the `EAGAIN` case. This is common practice in the Linux file systems when the page cache can not be invalided up front and falling back to the buffered path. In that case, the buffer is still stored in the page cache and will be reaped sometime later. This is an interesting idea though. @behlendorf what are you thoughts on this?

Reply:
With no input so far I think we can leave this to an add on PR.